File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.stream.IntStream;
import java.util.Arrays;

class Solution {
    int[] root;
    public int solution(int n, int[][] computers) {
        root=IntStream.range(0,n).toArray();
        for(int i=0;i<computers.length;i++){
            for(int j=i+1;j<computers[i].length;j++){
                if(computers[i][j]==1)
                    setRoot(i,j);
            }
        }
        
        for(int i=1;i<computers.length;i++) getRoot(i); //최종 업데이트
        
        return Arrays.stream(root).distinct().toArray().length;
    }
    
    private void setRoot(final int idx1, final int idx2){
        int root1=getRoot(idx1);
        int root2=getRoot(idx2);
        if(root1>root2){
            root[root1]=root2;
        }
        if(root1<root2){
            root[root2]=root1;
        }
        //동일하면 작업 없음
    }
    
    private int getRoot(final int idx){
        if(root[idx]==idx) return idx;
        return root[idx]=getRoot(root[idx]);
    }
}

/*
1부터 훑어 그거 만들기
만들고 전체 n 훑어 root 변경
distinct count
*/

File: 프로그래머스/3/43162. 네트워크/네트워크.js
function solution(n, computers) {
    const root=new Array(n);
    for(let i=0;i<root.length;i++)
        root[i]=i;
    for(let i=0;i<computers.length;i++)
        for(let j=i+1;j<computers[i].length;j++)
            if(computers[i][j]===1){
                //propagation
                let value;
                let idx;
                let prevValue;
                if(root[i]<root[j]){
                    value=root[i];
                    idx=j;
                }else{
                    value=root[j];
                    idx=i;
                }
                prevValue=idx; //while문 첫 진입을 위한 코드 맞추기
                do{
                    idx=prevValue;
                    prevValue=root[idx];
                    root[idx]=value;
                }while(prevValue!==idx);
            }
    for(let i=0;i<root.length;i++)
        root[i]=root[root[i]]; //root 최종 업데이트
    return new Set(root).size;
}